components/odh-notebook-controller/controllers/notebook_rbac.go

```diff
@@ -0,0 +1,166 @@
+	return &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rolebindingName,
+			Namespace: notebook.Namespace,
+			Labels: map[string]string{
+				"notebook-name": notebook.Name,
+			},
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      notebook.Name,
+				Namespace: notebook.Namespace,
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			Kind:     roleRefKind,
+			Name:     roleRefName,
+			APIGroup: "rbac.authorization.k8s.io",
+		},
+	}
+}
+
+// checkRoleExists checks if a Role or ClusterRole exists in the namespace.
+func (r *OpenshiftNotebookReconciler) checkRoleExists(ctx context.Context, roleRefKind, roleRefName, namespace string) (bool, error) {
+	if roleRefKind == "ClusterRole" {
+		// Check ClusterRole if roleRefKind is ClusterRole
+		clusterRole := &rbacv1.ClusterRole{}
+		err := r.Get(ctx, types.NamespacedName{Name: roleRefName}, clusterRole)
+		if err != nil {
+			if apierrs.IsNotFound(err) {
+				// ClusterRole not found
+				return false, nil
+			}
+			return false, err // Some other error occurred
+		}
+	} else {
+		// Check Role if roleRefKind is Role
+		role := &rbacv1.Role{}
+		err := r.Get(ctx, types.NamespacedName{Name: roleRefName, Namespace: namespace}, role)
+		if err != nil {
+			if apierrs.IsNotFound(err) {
+				// Role not found
+				return false, nil
+			}
+			return false, err // Some other error occurred
+		}
+	}
+	return true, nil // Role or ClusterRole exists
+}
+
+// Helper function to delete the RoleBinding if the notebooks deleted
+func (r *OpenshiftNotebookReconciler) deleteRoleBinding(ctx context.Context, rolebindingName, namespace string) error {
+	roleBinding := &rbacv1.RoleBinding{}
+	err := r.Get(ctx, types.NamespacedName{Name: rolebindingName, Namespace: namespace}, roleBinding)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			return nil // RoleBinding not found, nothing to delete
+		}
+		return err // Some other error occurred
+	}
+
+	// Delete the RoleBinding
+	return r.Delete(ctx, roleBinding)
+}
+
+// reconcileRoleBinding manages creation, update, and deletion of RoleBindings and ClusterRoleBindings
+func (r *OpenshiftNotebookReconciler) reconcileRoleBinding(
+	notebook *nbv1.Notebook, ctx context.Context, rolebindingName, roleRefKind, roleRefName string) error {
+
+	log := r.Log.WithValues("notebook", types.NamespacedName{Name: notebook.Name, Namespace: notebook.Namespace})
+
+	// Check if the Role or ClusterRole exists before proceeding
+	roleExists, err := r.checkRoleExists(ctx, roleRefKind, roleRefName, notebook.Namespace)
+	if err != nil {
+		log.Error(err, "Error checking if Role exists", "Role.Kind", roleRefKind, "Role.Name", roleRefName)
+		return err
+	}
+	if !roleExists {
+		return nil // Skip if dspa Role is not found on the namespace
+	}
+
+	// Create a new RoleBinding based on provided parameters
+	roleBinding := NewRoleBinding(notebook, rolebindingName, roleRefKind, roleRefName)
```

Comment:
Our odh-notebook-operator should probably `Own` the rolebindings now

https://github.com/opendatahub-io/kubeflow/blob/6e20d786ac6035a78bac135dc1aa246b3013e3f5/components/odh-notebook-controller/controllers/notebook_controller.go#L439-L447

Does it create with the correct labels, that is, the owner reference thingy?

Reply:
Good catch, I added under notebook_controller.
Now the rolebinding includes `ownerReferences`

```
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: elyra-pipelines-q1
  namespace: pipelines
  uid: edb9840d-9e48-4516-8974-5d14c47388d0
  resourceVersion: '23522948'
  creationTimestamp: '2024-10-30T15:11:33Z'
  labels:
    notebook-name: q1
  ownerReferences:
    - apiVersion: kubeflow.org/v1beta1
      kind: Notebook
      name: q1
      uid: 84fb0090-0258-48f7-86ff-6fb7d7272075
subjects:
  - kind: ServiceAccount
    name: q1
    namespace: pipelines
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: ds-pipeline-user-access-dspa
```